demos/meltdown_de.cc

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Demonstrates the Meltdown-DE on AMD.
+ * We exploit the fact that the data from the remainder on AMD is speculatively
+ * computed this way (before the #DE exception is raised and the remainder is
+ * zeroed):
+ * 0 % 0 = 0
+ * 1 % 0 = 2
+ * 2 % 0 = 2
+ * 3 % 0 = 3
+ * 4 % 0 = 2
+ * 5 % 0 = 2
+ * 6 % 0 = 3
+ * 7 % 0 = 3
+ * 8 % 0 = 2
+ * 9 % 0 = 2
+ * 10 % 0 = 2
+ * 11 % 0 = 2
+ * 12 % 0 = 3
+ * 13 % 0 = 3
+ * 14 % 0 = 3
+ * 15 % 0 = 3
+ * 16 % 0 = 2
+ * 7 times more 2
+ * 8 times 3
+ * 16 times 2
+ * 16 times 3
+ * 32 times 2
+ * 32 times 3
+ * etc.
+ * We use the second row, because it's the weirdest one, as the speculative
+ * remainder is bigger than both operands.
+ * Therefore we accomodate the private data to be stored in multiple strings
+ * where the secret data is on index 2 (first two indices contain dummy data).
+ * Then we speculatively access those unreachable characters in a loop.
+ **/
+
+#include "compiler_specifics.h"
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#if !SAFESIDE_IA32 && !SAFESIDE_X64
+#  error Unsupported architecture. AMD required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <signal.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+#include "utils.h"
+
+const char *public_data = "Hello, world!";
+
+// First two characters of each string are always dummy.
+const char *private_data[] = {
+  "XXI",
+  "XXt",
+  "XX'",
+  "XXs",
+  "XX ",
+  "XXa",
+  "XX ",
+  "XXs",
+  "XXe",
+  "XXc",
+  "XXr",
+  "XXe",
+  "XXt",
+  "XX!",
+  "XX!",
+  "XX!",
+};
+
+constexpr size_t kPrivateDataLength = 16;
```

Comment:
let's either compute this with `sizeof(private_data)/sizeof(*private_data)` or, maybe better, put the length definition first and use it in the array declaration: `const char *private_data[kPrivateDataLength] = ...`

Reply:
Done.